getSamples performance improvements

zagrodzki opened the pull request:
Benchmark results:
BenchmarkGetSamples/single_make-4         	   50000	     28906 ns/op
BenchmarkGetSamples/two_makes-4           	   50000	     37339 ns/op
BenchmarkGetSamples/fast_lookup_one_loop-4         	   30000	     50444 ns/op
BenchmarkGetSamples/fast_lookup_two_loops-4        	   50000	     35126 ns/op

PProf results from running a 10 second capture:
before: https://pastebin.com/vv8YBLiT
after: https://pastebin.com/JvxEmdJm

Time spent in getSamples for this run dropped from 17% of the runtime to <5% of the runtime.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zagrodzki/goscope/40)
<!-- Reviewable:end -->

zagrodzki committed:
Use a lookup table for samples, avoiding expensive floating point
calculation for every received sample later.

zagrodzki committed:
iterate channel by channel, this avoids the mod calculation in every
step.

zagrodzki committed:
Use numChan, not 2.

zagrodzki committed:
Test two allocs vs one alloc.
Test two nested loops vs single loop.

zagrodzki committed:
avoid expensive mod in single loop. didn't help much anyway.

zagrodzki committed:
Use a single make() to allocate a larger buffer, then use slices
to pass it around.

zagrodzki committed:
add missing license snippet

dtkaczyk commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[usb/hantek6022be/capture.go, line 68 at r2](https://reviewable.io:443/reviews/zagrodzki/goscope/40#-KsohKVzLwq91Mk4Kcii:-KsohKW-CBQelz7tCF3k:b-7h96mh) ([raw file](https://github.com/zagrodzki/goscope/blob/cb01b3ff92fac6c467437804127a6f9b1974d94d/usb/hantek6022be/capture.go#L68)):*
> ```Go
> 
> type captureParams struct {
> 	translateSample [2][256]scope.Voltage
> ```

We have a lot of mysterious numbers in the code. It makes it harder to understand what's going on. Maybe we could use constants with meaningful names instead? Not necessarily for all the existing code, but maybe for new commits?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->

zagrodzki commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[usb/hantek6022be/capture.go, line 68 at r2](https://reviewable.io:443/reviews/zagrodzki/goscope/40#-KsohKVzLwq91Mk4Kcii:-KspKpfeY9yPhEb9YIOG:b-1co825) ([raw file](https://github.com/zagrodzki/goscope/blob/cb01b3ff92fac6c467437804127a6f9b1974d94d/usb/hantek6022be/capture.go#L68)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

We have a lot of mysterious numbers in the code. It makes it harder to understand what's going on. Maybe we could use constants with meaningful names instead? Not necessarily for all the existing code, but maybe for new commits?
</blockquote></details>

these are not that mysterious. I can add a const for maxChans = 2 for the size of the outer array.
For the inner array the intention is "all possible byte values". "numBytes" might be confusing, since it typically indicates a length not  number of possible values. maxByte+1 (where maxByte is 255) would be correct, but ugly. Ideas?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->

zagrodzki commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[usb/hantek6022be/capture.go, line 68 at r2](https://reviewable.io:443/reviews/zagrodzki/goscope/40#-KsohKVzLwq91Mk4Kcii:-KspLS333OyJLSEmpcVZ:b3jctyo) ([raw file](https://github.com/zagrodzki/goscope/blob/cb01b3ff92fac6c467437804127a6f9b1974d94d/usb/hantek6022be/capture.go#L68)):*
<details><summary><i>Previously, zagrodzki wrote…</i></summary><blockquote>

these are not that mysterious. I can add a const for maxChans = 2 for the size of the outer array.
For the inner array the intention is "all possible byte values". "numBytes" might be confusing, since it typically indicates a length not  number of possible values. maxByte+1 (where maxByte is 255) would be correct, but ugly. Ideas?
</blockquote></details>

I could also just add a comment on translateSample field, that might actually be a better solution? the numbers are not interesting, but the meaning of the entire construct is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->

dtkaczyk commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[usb/hantek6022be/capture.go, line 68 at r2](https://reviewable.io:443/reviews/zagrodzki/goscope/40#-KsohKVzLwq91Mk4Kcii:-KsqzRaepnJTejhF_gtF:b-q4hf18) ([raw file](https://github.com/zagrodzki/goscope/blob/cb01b3ff92fac6c467437804127a6f9b1974d94d/usb/hantek6022be/capture.go#L68)):*
<details><summary><i>Previously, zagrodzki wrote…</i></summary><blockquote>

I could also just add a comment on translateSample field, that might actually be a better solution? the numbers are not interesting, but the meaning of the entire construct is.
</blockquote></details>

Short answer: I would add a comment (it's necessary) and maxChans = 2 (I vote for it, but I'll leave the decision to you). The inner number is not that bad.

Longer explanation: I realize the numbers are not mysterious for the author of them ;). They were mysterious for me. By that I mean: when I look at this declaration, I don't understand what is translateSample, or why its dimensions are what they are. To understand that, I had to analyze another piece of code that is in a different place (the one that uses translateSample). My feeling is that it's not a good situation, I imagine it should be fairly easy to at least have a vague idea what  this piece of code is for without searching in other places.

What bothers me more is that this is just an example, I believe in other files as well there are many similar situations. Of course if you strongly oppose any changes in this practice, say it now and I will remain silent about this issue forever :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->

zagrodzki committed:
Add maxChans const to use for array/slice sizes where data is stored
per channel. Comments for translateSamples. Iterate over range
instead of numerical indexes in sample translations, to avoid repeating
the size constants.

zagrodzki commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[usb/hantek6022be/capture.go, line 68 at r2](https://reviewable.io:443/reviews/zagrodzki/goscope/40#-KsohKVzLwq91Mk4Kcii:-KsrZv552VIFRGztmBSt:bbtp9st) ([raw file](https://github.com/zagrodzki/goscope/blob/cb01b3ff92fac6c467437804127a6f9b1974d94d/usb/hantek6022be/capture.go#L68)):*
<details><summary><i>Previously, dtkaczyk (Dominika Tkaczyk) wrote…</i></summary><blockquote>

Short answer: I would add a comment (it's necessary) and maxChans = 2 (I vote for it, but I'll leave the decision to you). The inner number is not that bad.

Longer explanation: I realize the numbers are not mysterious for the author of them ;). They were mysterious for me. By that I mean: when I look at this declaration, I don't understand what is translateSample, or why its dimensions are what they are. To understand that, I had to analyze another piece of code that is in a different place (the one that uses translateSample). My feeling is that it's not a good situation, I imagine it should be fairly easy to at least have a vague idea what  this piece of code is for without searching in other places.

What bothers me more is that this is just an example, I believe in other files as well there are many similar situations. Of course if you strongly oppose any changes in this practice, say it now and I will remain silent about this issue forever :)
</blockquote></details>

Done.

I'm not opposing any of that, of course. The entire purpose of the peer review is to make sure the code is readable not just to the person who wrote it. You should not need to read multiple places, jumping all around the code, to understand what the code does. Typically the variable or structure declaration should have enough information.

Feel free to point out any other places where you think  the purpose of the code or it's meaning is not clear. It's difficult for me to see that for myself immediately, since everything seems obvious just after you write it...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->

dtkaczyk commented:
Reviewed 1 of 1 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zagrodzki/goscope/40)*
<!-- Sent from Reviewable.io -->
